### PR TITLE
Add URL Trim Query UDF

### PR DESCRIPTION
**Pull request by 1987yama3**
I implement a BigQuery UDF to trim URL query parameter. 
For example, in Google Analytics 4 BigQuery, the "page_location" parameter store full query parameter. 
However, in the context of web analytics, there is a need to exclude some query parameter for analytics.

Excluding them with Standard Function is complicated, so I implemented UDF.

Thanks

### CODE REVIEW

File: udfs/community/test_cases.js [hunk 1]
```diff
@@ -452,6 +452,43 @@ generate_udf_test("url_param", [
         expected_output: `NULL`
     },
 ]);
+generate_udf_test("url_trim_query", [
```
Comment (danieldeleo) [inline udfs/community/test_cases.js:455]: Add test cases for NULL inputs and a test case for url which contains keys but no hash

**Review by danieldeleo — CHANGES_REQUESTED** [review]
Thanks for creating this PR. I think it would be better to keep UDFs in SQL syntax if possible since they'll be more performant. I've created an example below which I think should provide the same functionality as your JS UDF. 


`CREATE TEMP FUNCTION url_trim_keys(url STRING, keys_to_trim ARRAY<STRING>) AS ((

SELECT 
    url_base || IF(keys IS NULL, '', '?' || keys) || IF(url_hash IS NULL, '', '#' || url_hash)
FROM(
    SELECT 
        url_base,
        url_hash,
        STRING_AGG(
            IF(LOWER(SPLIT(url_key, '=')[SAFE_OFFSET(0)]) NOT IN(SELECT LOWER(key) FROM UNNEST(keys_to_trim) AS key),
                url_key,
                NULL),
            '&') AS keys
    FROM(
        SELECT
            SPLIT(url, '?')[SAFE_OFFSET(0)] AS url_base,
            SPLIT(url, '#')[SAFE_OFFSET(1)] AS url_hash,
            IFNULL(SPLIT(SPLIT(SPLIT(url, '?')[SAFE_OFFSET(1)], '#')[SAFE_OFFSET(0)], '&'), [CAST(NULL AS STRING)]) AS url_keys
    ), UNNEST(url_keys) AS url_key
    GROUP BY 1,2)
));

SELECT url_trim("https://www.example.com/abc/index.html?id=12345&utm_id=abc123&hello=world", ["utm_id", "id"])
UNION ALL
SELECT url_trim("https://www.example.com/abc/index.html?id=12345&utm_id=abc123",["utm_id"])
UNION ALL        
SELECT url_trim("https://www.example.com/abc/index.html?id=12345&utm_id=abc123",["utm_id", "id"])
UNION ALL       
SELECT url_trim( "https://www.example.com/abc/index.html?id=12345&utm_id=abc123#hash", ["utm_id"])
UNION ALL
SELECT url_trim("https://www.example.com/abc/index.html?id=12345&utm_id=abc123#hash", ["utm_id", "id"])
UNION ALL       
SELECT url_trim("https://www.example.com/abc/index.html", ["gclid"])`

**Comment by 1987yama3** [comment]
@danieldeleo Thanks for your comment. It is nice code! I change to SQL Syntax from JS. Please check this.

File: udfs/community/url_trim_query.sqlx [hunk 2]
```diff
@@ -0,0 +1,37 @@
+config { hasOutput: true }
+/*
+ * Copyright 2019 Google LLC
```
Comment (danieldeleo) [inline udfs/community/url_trim_query.sqlx:3]:
```suggestion
 * Copyright 2022 Google LLC
```

File: udfs/community/test_cases.js [hunk 3]
```diff
@@ -452,6 +452,72 @@ generate_udf_test("url_param", [
...
+        expected_output: `"https://www.example.com/abc/index.html?id=12345"`
+    },
+    {
+        inputs: [
+          `"https://www.example.com/abc/index.html?id=12345&utm_id=abc123"`,
+          `["utm_id", "id"]`
+        ],
+        expected_output: `"https://www.example.com/abc/index.html"`
+    },
+
```
Comment (danieldeleo) [inline udfs/community/test_cases.js:498]:
```suggestion
```

File: udfs/community/test_cases.js [hunk 4]
```diff
@@ -452,6 +452,72 @@ generate_udf_test("url_param", [
...
+        expected_output: `"https://www.example.com/abc/index.html#hash"`
+    },
+    {
+        inputs: [
+            `"https://www.example.com/abc/index.html"`,
+            `["gclid"]`
+        ],
+        expected_output: `"https://www.example.com/abc/index.html"`
+    },
+])
```
Comment (danieldeleo) [inline udfs/community/test_cases.js:520]:
```suggestion
]);
```

File: udfs/community/test_cases.js [hunk 5]
```diff
@@ -452,6 +452,72 @@ generate_udf_test("url_param", [
         expected_output: `NULL`
     },
 ]);
+generate_udf_test("url_trim_query", [
+    {
+        inputs: [
+            `NULL`,
+            `["hellp"]`
```
Comment (danieldeleo) [inline udfs/community/test_cases.js:459]:
```suggestion
            `["hello"]`
```

File: udfs/community/README.md [hunk 6]
```diff
@@ -748,6 +749,30 @@ results:
 | facebook.com | NULL        | NULL             | NULL | rpc  |
 
 
+### [url_trim_query(urlString STRING, keys ARRAY<STRING>)](url_trim_query.sqlx)
```
Comment (danieldeleo) [inline udfs/community/README.md:752]:
```suggestion
### [url_trim_query(url STRING, keys_to_trim ARRAY<STRING>)](url_trim_query.sqlx)
```

File: udfs/community/README.md [hunk 7]
```diff
@@ -748,6 +749,30 @@ results:
...
+
+Returns trimed query URL from input URL. The remove query is specified arguments.
```
Comment (danieldeleo) [inline udfs/community/README.md:754]:
```suggestion
Returns a URL with specified keys removed from the
[URL's query component](https://en.wikipedia.org/wiki/Query_string). 
The keys to be removed are provided as an ARRAY<STRING> input argument.
```

File: udfs/community/url_trim_query.sqlx (hunk 2 above)
Comment (1987yama3) [reply udfs/community/url_trim_query.sqlx]: Thanks

**Comment by 1987yama3** [comment]
@danieldeleo Thanks. I fixed according to the code review. Please check it.

**Review by danieldeleo — APPROVED** [review]
Thanks for the quick turnaround!
